keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/diff/DualNumber.java

```diff
@@ -350,4 +351,84 @@ private DoubleTensor pluckFromPartial(DoubleTensor from, int... indices) {
+        } else {
+            return pluckFromMixedPredicate(predicate, this, els);
+        }
+    }
+
+    private DualNumber pluckFromMixedPredicate(BooleanTensor predicate, DualNumber thn, DualNumber els) {
+        int[] thenShape = thn.value.getShape();
+        double[] flatPredicate = predicate.asFlatDoubleArray();
+        Map<Long, List<DoubleTensor>> ifAndElseDualNumbers = new HashMap<>();
+
+        for (int i = 0; i < flatPredicate.length; i++) {
+            boolean condition = flatPredicate[i] == 1.0;
+            if (condition) {
+                pluckFromThenAndElse(thn, els, ifAndElseDualNumbers, thenShape, i);
+            } else {
+                pluckFromThenAndElse(els, thn, ifAndElseDualNumbers, thenShape, i);
+            }
+        }
+
+        Map<Long, DoubleTensor> newPartials = concatenatePluckedPartials(ifAndElseDualNumbers, thn, els);
+        return new DualNumber(predicate.setDoubleIf(thn.value, els.value), newPartials);
+    }
+
+    private void pluckFromThenAndElse(DualNumber primary, DualNumber secondary, Map<Long, List<DoubleTensor>> partials, int[] shape, int index) {
+        int[] currentIndex = TensorShape.getShapeIndices(shape, TensorShape.getRowFirstStride(shape), index);
+        DualNumber primaryDualNumber = primary.pluck(currentIndex);
+        DualNumber secondaryDualNumber = secondary.pluck(currentIndex);
+        Map<Long, DoubleTensor> primaryDualsMap = primaryDualNumber.getPartialDerivatives().asMap();
+        Map<Long, DoubleTensor> secondaryWithPrimaryRemoved = removePrimaryFromSecondaryAndZero(primaryDualNumber, secondaryDualNumber);
+
+        addToMap(partials, primaryDualsMap, secondaryWithPrimaryRemoved);
+    }
+
+    private Map<Long, DoubleTensor> concatenatePluckedPartials(Map<Long, List<DoubleTensor>> ifAndElseDualNumbers, DualNumber thn, DualNumber els) {
+        Map<Long, DoubleTensor> newPartials = new HashMap<>();
+
+        for (Map.Entry<Long, List<DoubleTensor>> entry : ifAndElseDualNumbers.entrySet()) {
+            List<DoubleTensor> value = entry.getValue();
+            DoubleTensor primaryTensor = value.remove(0);
+            DoubleTensor[] partialsToConcat = new DoubleTensor[value.size()];
+            DoubleTensor concatenatedPartials = primaryTensor.concat(0, value.toArray(partialsToConcat));
+            int[] originalPartialShape = thn.getPartialDerivatives().asMap().containsKey(entry.getKey()) ?
+                thn.getPartialDerivatives().withRespectTo(entry.getKey()).getShape() :
+                els.getPartialDerivatives().withRespectTo(entry.getKey()).getShape();
+            newPartials.put(entry.getKey(), concatenatedPartials.reshape(originalPartialShape));
+        }
+
+        return newPartials;
+    }
+
+    private Map<Long, List<DoubleTensor>> addToMap(Map<Long, List<DoubleTensor>> toBeConcatted, Map<Long, DoubleTensor> a, Map<Long, DoubleTensor> b) {
+        for (Map.Entry<Long, DoubleTensor> entry : a.entrySet()) {
```

Comment:
you could probably combine the 2 loops into 1 if you take the union of the entrySets

Reply:
Nice - changed to a union